cardcomm/pkcs11/src/common/libtomcrypt/tomcrypt_hash.h

```diff
@@ -29,49 +29,101 @@ extern "C"
 
 /* ---- HASH FUNCTIONS ---- */
 #ifdef USE_SHA512
-	struct sha512_state
+#ifdef __LP64__
+struct sha512_state
+{
+    ulong64 length;
+    ulong64 state[8];
+    unsigned long curlen;
+    unsigned char buf[128];
+};
+#else
+    struct sha512_state
 	{
-		ulong64 length, state[8];
+        ulong32 length;
+        ulong32 state[8];
```

Comment:
No. You're dropping bits here, for a SHA512 algorithm. That's a terrible idea.

Did you run the hashing code, and did it return the same value?